[#40] Previews (#96)
github/actionview-component: README.md

end
```

### Previewing Components

> Action View Component previews provide a way to see how components look by visiting a special URL that renders them.

Reviewer: ```suggestion
`ActionView::Component::Preview`s provide a way to see how components look by visiting a special URL that renders them.
```

> In the previous example, the preview class for `TestComponent` should be named `TestComponentPreview` and located in `test/components/previews/test_component_preview.rb`.

Reviewer: ```suggestion
In the previous example, the preview class for `TestComponent` would be called `TestComponentPreview` and located in `test/components/previews/test_component_preview.rb`.
```

To see the preview of the component with a given title, implement a method that returns the locals (as a hash) to be passed to the component.
The method can be named after what you want to preview given the locals you use, and you can define many examples (methods) as you want:

```ruby
# test/components/previews/test_component_preview.rb

class TestComponentPreview < ActionView::Component::Preview
  def with_default_title

>     {

Reviewer: I'm curious as to the decision to go with a hash here, instead of something that more closely mimics the way components are rendered in views.

For `ActionMailer::Preview`s, the instance methods contain a direct call to the mailer in question: https://guides.rubyonrails.org/action_mailer_basics.html#previewing-emails. Ideally, we'd do the same here. 

Perhaps we could define our own `render` method on `ActionView::Component::Preview` that uses our `render_inline` test helper under the hood?

Author: Sounds good.

Do you think something like this would work?
```ruby
class TestComponentPreview < ActionView::Component::Preview
  def with_default_title
    render(title: "Default title", subtitle: "Some subtitle")
  end
end
```

And `ActionView::Component::Preview#render` would be an instance method of the preview object, right?

Reviewer: I'm thinking we should be even more explicit:

`render(TestComponent, title: "Default title", subtitle: "Some subtitle")`

The benefit of this is that the previews would be examples that could then be copied and pasted directly into views.

Author: I see, this is great

I'll update

      title: "Test component default"
    }
  end

  def with_long_title
    {
      title: "This is a really long title to see how the component renders this"
    }
  end
end
```

Then the previews will be available in <http://localhost:3000/rails/components/test_component/with_default_title>
and <http://localhost:3000/rails/components/test_component/with_long_title>.

> To specify the stylesheets and javascript used by your previews you must create a dummy layout file in `test/components/previews/layouts/`.

Reviewer: Hmm. I'm wondering if it might be possible to reuse at least part of the application layout here. I'm worried that having a separate layout file for this would cause the previews to be inaccurate.

Reviewer: At the very least, perhaps we could define a default layout that includes the default application styles and JS? That way developers could override that layout if necessary, but wouldn't have to create one. #conventionoverconfiguration

Author: Yep, great idea! 

So to sum up:
1. Create a default layout that will be used for rendering the preview examples
2. Developers should be able to override this layout

Author: Instead of creating a default layout inside the gem, do you think it's fine if we just use the application's layout? https://github.com/github/actionview-component/pull/96/commits/7df33aff785906097ad5fbaa0efa4dac381adac0#diff-67f888315429841b717b9aeef54e71a6R48

Developers will be able to set any other layout they want, but layouts must live inside their app. Makes sense?

Author: What do you think if we actually use the application's layout? https://github.com/github/actionview-component/pull/96/commits/7df33aff785906097ad5fbaa0efa4dac381adac0#diff-67f888315429841b717b9aeef54e71a6R48 So that way we don't need to create any layout on the gem's side. 

Developers will still be able to override this layout, using any of their application's layouts.

Reviewer: I think it's worth a shot! If we go that route, do we still need to us iframes?

Author: Yes, because iframes will contain the component html inside the application's layout. And those iframes will live inside a view that uses the rails' application layout.

Does it make sense to you? Or do you have any other approach in mind?

Reviewer: I guess I'm curious as to whether we need the iframes at all if we are just using the application layout.

I guess that would only work if you weren't using a different layout. Perhaps we could remove the custom layout functionality for now and see if anyone asks for it?

Author: Oh maybe we are not talking about the same thing. I forgot to update the readme. 

There are no custom layouts for component previews, developers can only use their application's layouts. By default previews will use the `app/views/layouts/application.html.erb` layout. 

But since the previews are displayed from a rails namespace, they are served [within a layout that comes from rails itself](https://github.com/rails/rails/blob/master/railties/lib/rails/templates/layouts/application.html.erb).

That's why we use iframes, to avoid having that layout when rendering the component. And also to avoid using the application's layout in the `rails/components` and `rails/components/:component_name` routes

I'll update the readme tonight

Reviewer: Gotcha. Based on my comment below about staying consistent with ActionMailer::Preview, I think this is just fine for now. 😄 👍

```erb
<!-- test/components/previews/layouts/application.html.erb -->

<!DOCTYPE html>
<html>
<head>
  <title>Dummy Layout</title>
  <%= stylesheet_link_tag "application" %>
</head>
<body>
  <%= yield %>
</body>
</html>
```

All previews will use `test/components/previews/layouts/application.*` as the default layout. You can set a different layout by calling `set_layout` in your preview class.

```ruby
# test/components/previews/test_component_preview.rb
class TestComponentPreview < ActionView::Component::Preview

>   set_layout :admin

Reviewer: Might we call this `layout`, to match https://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-layouts?

Author: Good call

end
```

### Previewing Components
ActionView::Component::Preview`s provide a way to see how components look by visiting a special URL that renders them.
In the previous example, the preview class for `TestComponent` would be called `TestComponentPreview` and located in `test/components/previews/test_component_preview.rb`.

> To see the preview of the component with a given title, implement a method that calls the ActionView's render method.

Reviewer: ```suggestion
To see the preview of the component with a given title, implement a method that renders the component.
```

> The method can be named after what you want to preview given the locals you use, and you can define many examples (methods) as you want:

Reviewer: ```suggestion
You can define as many examples as you want:
```

```ruby
# test/components/previews/test_component_preview.rb

class TestComponentPreview < ActionView::Component::Preview
  def with_default_title
    render(TestComponent, title: "Test component default")
  end

  def with_long_title
    render(TestComponent, title: "This is a really long title to see how the component renders this")
  end
end
```

> Then the previews will be available in <http://localhost:3000/rails/components/test_component/with_default_title>

Reviewer: ```suggestion
The previews will be available in <http://localhost:3000/rails/components/test_component/with_default_title>
```

and <http://localhost:3000/rails/components/test_component/with_long_title>.

> Previews will use your application's layout `app/views/layouts/application`. But you can also use any other layout from your app to render the preview.

Reviewer: ```suggestion
Previews use the application layout by default, but you can also use other layouts from your app:
```

```ruby
# test/components/previews/test_component_preview.rb

class TestComponentPreview < ActionView::Component::Preview
  layout "admin"

  ...
end
```

By default, the preview classes live in `test/components/previews`.
This can be configured using the `preview_path` option.

> For example, if you want to change it to `lib/component_previews`, you can configure it in config/application.rb:

Reviewer: ```suggestion
For example, if you want to use `lib/component_previews`, set the following in `config/application.rb`:
```
